// stats/index.js
  function getMaxIndex(array, sortType) {
    if (!array && !array.isArray(array)) return -1
    let maxIndex = 0

    for (let index = 0; index < array.length; index++) {
        if (array[index] && array[index + 1]) {
            if (sortType(array[index], array[index + 1]) < 0) {
                maxIndex = index + 1
            }
        }

    }
    return maxIndex
}

  function getMaxElement(array, sortType) {
    let index = getMaxIndex(array, sortType)
    return index === -1 ? null : array[index]
}

  function getMinIndex(array, sortType) {
    if (!array && !array.isArray(array)) return -1
    let maxIndex = 0

    for (let index = 0; index < array.length; index++) { 
        if (array[index] && array[index + 1]) {
            if (sortType(array[index], array[index + 1]) > 0) {
                maxIndex = index + 1
            }
        }

    }
    return maxIndex
}

  function getMinElement(array, sortType) {
    let index = getMinIndex(array, sortType)
    return index === -1 ? null : array[index]
}
 
  function getMedianIndex(input, comparator) {
    if (input.length === 0) {
        return -1;
    }
    var data = input.slice().sort(comparator);
    return input.indexOf(data[Math.floor(data.length / 2)]);
}

  function getMedianElement(input, comparator) {
    var index = getMedianIndex(input, comparator);
    return index === -1 ? null : input[index];
}
  function getAverageValue(input, getValue) {
    if (input.length === 0) {
        return null;
    }
    return input.reduce(
        function (result, item) {
            return result + getValue(item);
        },
        0
    ) / input.length;

}

module.exports = {
    getMaxIndex,
    getMaxElement,
    getMinIndex,
    getMinElement,
    getMedianIndex,
    getMedianElement,
    getAverageValue
};
